zebra-chain/src/value_balance.rs

```diff
@@ -0,0 +1,297 @@
+    pub fn sapling_amount(&self) -> Amount<C> {
+        self.sapling
+    }
+
+    /// Insert a sapling value balance into a given [`ValueBalance`]
+    /// leaving the other values untouched.
+    pub fn set_sapling_value_balance(&mut self, sapling_value_balance: ValueBalance<C>) -> &Self {
+        self.sapling = sapling_value_balance.sapling;
+        self
+    }
+
+    /// Get the orchard amount from the [`ValueBalance`].
+    pub fn orchard_amount(&self) -> Amount<C> {
+        self.orchard
+    }
+
+    /// Insert an orchard value balance into a given [`ValueBalance`]
+    /// leaving the other values untouched.
+    pub fn set_orchard_value_balance(&mut self, orchard_value_balance: ValueBalance<C>) -> &Self {
+        self.orchard = orchard_value_balance.orchard;
+        self
+    }
+}
+
+impl<C> Default for ValueBalance<C>
+where
+    C: Constraint + Copy,
+{
+    fn default() -> Self {
+        let zero = Amount::zero();
+        Self {
+            transparent: zero,
+            sprout: zero,
+            sapling: zero,
+            orchard: zero,
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug, Clone, PartialEq)]
+/// Errors that can be returned when validating a [`ValueBalance`].
+pub enum ValueBalanceError {
+    #[error("value balance contains invalid amounts")]
+    /// Any error related to [`Amount`]s inside the [`ValueBalance`]
+    AmountError(#[from] Error),
+}
+
+impl<C> std::ops::Add for ValueBalance<C>
+where
+    C: Constraint,
+{
+    type Output = Result<ValueBalance<C>, ValueBalanceError>;
+    fn add(self, rhs: ValueBalance<C>) -> Self::Output {
+        Ok(ValueBalance::<C> {
+            transparent: (self.transparent + rhs.transparent)?,
+            sprout: (self.sprout + rhs.sprout)?,
+            sapling: (self.sapling + rhs.sapling)?,
+            orchard: (self.orchard + rhs.orchard)?,
+        })
+    }
+}
+impl<C> std::ops::Add<ValueBalance<C>> for Result<ValueBalance<C>, ValueBalanceError>
+where
+    C: Constraint,
+{
+    type Output = Result<ValueBalance<C>, ValueBalanceError>;
+    fn add(self, rhs: ValueBalance<C>) -> Self::Output {
+        self? + rhs
+    }
+}
+
+impl<C> std::ops::Sub for ValueBalance<C>
+where
+    C: Constraint,
+{
+    type Output = Result<ValueBalance<C>, ValueBalanceError>;
+    fn sub(self, rhs: ValueBalance<C>) -> Self::Output {
+        Ok(ValueBalance::<C> {
+            transparent: (self.transparent - rhs.transparent)?,
+            sprout: (self.sprout - rhs.sprout)?,
+            sapling: (self.sapling - rhs.sapling)?,
+            orchard: (self.orchard - rhs.orchard)?,
+        })
+    }
+}
+impl<C> std::ops::Sub<ValueBalance<C>> for Result<ValueBalance<C>, ValueBalanceError>
+where
+    C: Constraint,
+{
+    type Output = Result<ValueBalance<C>, ValueBalanceError>;
+    fn sub(self, rhs: ValueBalance<C>) -> Self::Output {
+        self? - rhs
+    }
+}
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest::prelude::*;
```

Comment:
Module structure: Can you move the `Arbitrary` impls and tests to submodules?

Reply:
https://github.com/ZcashFoundation/zebra/pull/2505/commits/70ec9c266a77f57aaf6f4dfcc1ec2d3f2d9eeb5b